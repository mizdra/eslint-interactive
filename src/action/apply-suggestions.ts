import { access, mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import type { Remote } from 'comlink';
import type { ESLint } from 'eslint';
import { promptToInputReuseFilterScript } from '../cli/prompt.js';
import { fixingSpinner } from '../cli/spinner.js';
import type { Undo } from '../core.js';
import type { SerializableCore } from '../core-worker.js';
import {
  editFileWithEditor,
  generateExampleFilterScriptFilePath,
  generateFilterScriptFilePath,
} from '../util/filter-script.js';

export async function doApplySuggestionsAction(
  core: Remote<SerializableCore>,
  results: ESLint.LintResult[],
  selectedRuleIds: string[],
): Promise<Undo> {
  const exampleScript = await readFile(generateExampleFilterScriptFilePath(), 'utf8');
  const filterScriptFilePath = generateFilterScriptFilePath(selectedRuleIds);
  const isFilterScriptExist = await access(filterScriptFilePath)
    .then(() => true)
    .catch(() => false);
  if (isFilterScriptExist) {
    const reuseFilterScript = await promptToInputReuseFilterScript();
    if (!reuseFilterScript) {
      await writeFile(filterScriptFilePath, exampleScript);
    }
  } else {
    await mkdir(dirname(filterScriptFilePath), { recursive: true }); // Create the directory because it might not exist
    await writeFile(filterScriptFilePath, exampleScript);
  }
  console.log('Opening editor...');

  const filterScript = await editFileWithEditor(filterScriptFilePath);
  const undo = await fixingSpinner(async () => core.applySuggestions(results, selectedRuleIds, filterScript));
  return undo;
}
