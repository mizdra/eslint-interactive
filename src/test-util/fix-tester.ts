import type { Linter, Rule } from 'eslint';
import { verifyAndFix } from '../eslint/linter.js';
import { LegacyESLint } from '../eslint/use-at-your-own-risk.js';
import type { FixContext } from '../fix/index.js';
import { plugin } from '../plugin.js';

const DEFAULT_FILENAME = 'test.js';

/**
 * The type representing the test case.
 */
type TestCase<FixArgs> = {
  /**
   * The filename of the code.
   */
  filename?: string;
  /**
   * The code to be fixed.
   */
  code: string | string[];
  /**
   * The rules to fix.
   */
  rules: Partial<Linter.RulesRecord>;
  /**
   * The arguments to pass to the fix function.
   */
  args?: FixArgs;
};

type TestResult = string | null;

/**
 * The test utility for the fix.
 */
export class FixTester<FixArgs> {
  private fixCreator: (context: FixContext, args: FixArgs) => Rule.Fix[];
  private defaultFixArgs: FixArgs;
  private defaultLinterConfig: Linter.LegacyConfig;
  constructor(
    fixCreator: (context: FixContext, args: FixArgs) => Rule.Fix[],
    defaultFixArgs: FixArgs,
    defaultLinterConfig: Linter.LegacyConfig,
  ) {
    this.fixCreator = fixCreator;
    this.defaultFixArgs = defaultFixArgs;
    this.defaultLinterConfig = defaultLinterConfig;
  }
  /**
   * Test the fix.
   * @param testCase The test case.
   * @returns The fixed code. If the fix skipped, null is returned.
   */
  async test(testCase: TestCase<FixArgs>): Promise<TestResult> {
    const code = Array.isArray(testCase.code) ? testCase.code.join('\n') : testCase.code;

    const filePath = testCase.filename ?? DEFAULT_FILENAME;

    const ruleIdsToFix = Object.keys(testCase.rules);
    const eslint = new LegacyESLint({
      useEslintrc: false,
      plugins: {
        'eslint-interactive': plugin,
      },
      overrideConfig: {
        ...this.defaultLinterConfig,
        plugins: ['eslint-interactive', ...(this.defaultLinterConfig.plugins ?? [])],
        rules: {
          ...this.defaultLinterConfig.rules,
          ...testCase.rules,
          'eslint-interactive/source-code-snatcher': 'error',
        },
      },
    });
    const fixedResult = await verifyAndFix(eslint, code, filePath, ruleIdsToFix, (context) =>
      this.fixCreator(context, testCase.args ?? this.defaultFixArgs),
    );

    if (fixedResult.fixed) {
      return fixedResult.output;
    } else {
      return null;
    }
  }
}
