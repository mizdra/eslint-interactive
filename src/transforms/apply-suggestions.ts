import { Linter, Rule } from 'eslint';
import { TransformContext } from '../types.js';

export type SuggestionFilter = (
  suggestions: Linter.LintSuggestion[],
  message: Linter.LintMessage,
) => Linter.LintSuggestion | null | undefined;

export type TransformToApplySuggestionsArgs = {
  filter: SuggestionFilter;
};

function getApplicableSuggestion(message: Linter.LintMessage, filter: SuggestionFilter): Linter.LintSuggestion | null {
  if (!message.suggestions || message.suggestions.length === 0) return null;
  const suggestion = filter(message.suggestions, message);
  return suggestion ?? null;
}

function generateFixPerMessage(
  _context: TransformContext,
  filter: SuggestionFilter,
  message: Linter.LintMessage,
): Rule.Fix | null {
  const suggestion = getApplicableSuggestion(message, filter);
  if (!suggestion) return null;
  return suggestion.fix;
}

/**
 * Create transform to apply suggestions.
 */
export function createTransformToApplySuggestions(
  context: TransformContext,
  args: TransformToApplySuggestionsArgs,
): Rule.Fix[] {
  const fixes = [];
  for (const message of context.messages) {
    const fix = generateFixPerMessage(context, args.filter, message);
    if (fix) fixes.push(fix);
  }
  return fixes;
}
