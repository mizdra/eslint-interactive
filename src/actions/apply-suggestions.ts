import { access, mkdir, readFile, writeFile } from 'fs/promises';
import { dirname } from 'path';
import chalk from 'chalk';
import { ESLint } from 'eslint';
import ora from 'ora';
import { promptToInputReuseFilterScript } from '../cli/prompt.js';
import { Core } from '../core.js';
import { SuggestionFilter } from '../transforms/apply-suggestions.js';
import {
  editFileWithEditor,
  generateExampleFilterScriptFilePath,
  generateFilterScriptFilePath,
} from '../util/filter-script.js';

export async function doApplySuggestionsAction(
  core: Core,
  results: ESLint.LintResult[],
  selectedRuleIds: string[],
): Promise<void> {
  const exampleScript = await readFile(generateExampleFilterScriptFilePath(), 'utf8');
  const filterScriptFilePath = generateFilterScriptFilePath(selectedRuleIds);
  const isFilterScriptExist = await access(filterScriptFilePath)
    .then(() => true)
    .catch(() => false);
  if (isFilterScriptExist) {
    const reuseFilterScript = await promptToInputReuseFilterScript();
    if (!reuseFilterScript) {
      await writeFile(filterScriptFilePath, exampleScript);
    }
  } else {
    // ディレクトリがない可能性を考慮して作成しておく
    await mkdir(dirname(filterScriptFilePath), { recursive: true });
    await writeFile(filterScriptFilePath, exampleScript);
  }
  console.log('Opening editor...');

  const filterScript = await editFileWithEditor(filterScriptFilePath);
  const filter = eval(filterScript) as SuggestionFilter;
  const fixingSpinner = ora('Applying suggestion...').start();
  await core.applySuggestions(results, selectedRuleIds, filter);
  fixingSpinner.succeed(chalk.bold('Applying suggestion was successful.'));
}
