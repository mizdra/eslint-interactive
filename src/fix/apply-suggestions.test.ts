import { basename } from 'node:path';
import { describe, expect, test } from 'vitest';
import { FixTester } from '../test-util/fix-tester.js';
import { createFixToApplySuggestions, FixToApplySuggestionsArgs } from './apply-suggestions.js';

const tester = new FixTester<FixToApplySuggestionsArgs>(
  createFixToApplySuggestions,
  { filter: (suggestions) => suggestions[0] },
  { parserOptions: { ecmaVersion: 2020, ecmaFeatures: { jsx: true } } },
);

describe('apply-suggestions', () => {
  test('basic', async () => {
    expect(
      await tester.test({
        code: 'a = a + 1;',
        rules: { 'eslint-interactive/prefer-addition-shorthand': 'error' },
        args: { filter: (suggestions) => suggestions[0] },
      }),
    ).toMatchInlineSnapshot(`"a += 1;"`);
  });
  test('一度に複数の suggestion を適用できる', async () => {
    expect(
      await tester.test({
        code: ['a = a + 1;', 'b = b + 1;'],
        rules: { 'eslint-interactive/prefer-addition-shorthand': 'error' },
        args: { filter: (suggestions) => suggestions[0] },
      }),
    ).toMatchInlineSnapshot(`
      "a += 1;
      b += 1;"
    `);
  });
  test('一度に複数の rule の suggestion を適用できる', async () => {
    expect(
      await tester.test({
        code: ['a = a + 1;', 'if (!key in object) {}'],
        rules: { 'eslint-interactive/prefer-addition-shorthand': 'error', 'no-unsafe-negation': 'error' },
        args: { filter: (suggestions) => suggestions[0] },
      }),
    ).toMatchInlineSnapshot(`
      "a += 1;
      if (!(key in object)) {}"
    `);
  });
  test('1 つの行に複数の suggestion があっても全ての suggestion が適用できる', async () => {
    expect(
      await tester.test({
        code: ['a = a + 1; b = b + 1;'],
        rules: { 'eslint-interactive/prefer-addition-shorthand': 'error' },
        args: { filter: (suggestions) => suggestions[0] },
      }),
    ).toMatchInlineSnapshot(`"a += 1; b += 1;"`);
  });
  test('filter には suggestions, message が渡ってくる', async () => {
    await tester.test({
      code: ['a = a + 1;'],
      rules: { 'eslint-interactive/prefer-addition-shorthand': 'error' },
      args: {
        filter: (suggestions, message, context) => {
          expect({
            suggestions,
            message,
            context: {
              ...context,
              // Use only basename, because the path will change depending on the environment.
              filename: basename(context.filename),
              // Take a snapshot of only part of it because `sourceCode` is huge
              sourceCode: { text: context.sourceCode.text },
            },
          }).toMatchSnapshot();
          return null;
        },
      },
    });
  });
  test('suggestion がない場合は何もしない', async () => {
    expect(
      await tester.test({
        code: 'a = a + 1;',
        rules: { semi: 'error' },
        args: { filter: (suggestions) => suggestions[0] },
      }),
    ).toMatchInlineSnapshot(`null`);
  });
  test('filter から null もしくは undefined を返すと、suggestion は適用されない', async () => {
    expect(
      await tester.test({
        code: 'a = a + 1;',
        rules: { 'eslint-interactive/prefer-addition-shorthand': 'error' },
        args: { filter: (_suggestions) => (Math.random() < 0.5 ? null : undefined) },
      }),
    ).toMatchInlineSnapshot(`null`);
  });
});
