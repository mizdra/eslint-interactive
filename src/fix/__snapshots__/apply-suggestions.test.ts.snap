// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apply-suggestions > filter には suggestions, message が渡ってくる 1`] = `
{
  "context": {
    "filename": "test.js",
    "fixer": {
      "insertTextAfter": [Function],
      "insertTextAfterRange": [Function],
      "insertTextBefore": [Function],
      "insertTextBeforeRange": [Function],
      "remove": [Function],
      "removeRange": [Function],
      "replaceText": [Function],
      "replaceTextRange": [Function],
    },
    "messages": [
      {
        "column": 1,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "The addition method is redundant.",
        "nodeType": "AssignmentExpression",
        "ruleId": "eslint-interactive/prefer-addition-shorthand",
        "severity": 2,
        "suggestions": [
          {
            "desc": "Use \`val += 1\` instead.",
            "fix": {
              "range": [
                0,
                9,
              ],
              "text": "a += 1",
            },
          },
          {
            "desc": "Use \`val++\` instead.",
            "fix": {
              "range": [
                0,
                9,
              ],
              "text": "a++",
            },
          },
          {
            "desc": "Use \`++val\` instead.",
            "fix": {
              "range": [
                0,
                9,
              ],
              "text": "++a",
            },
          },
        ],
      },
    ],
    "ruleIds": [
      "eslint-interactive/prefer-addition-shorthand",
    ],
    "sourceCode": {
      "text": "a = a + 1;",
    },
  },
  "message": {
    "column": 1,
    "endColumn": 10,
    "endLine": 1,
    "line": 1,
    "message": "The addition method is redundant.",
    "nodeType": "AssignmentExpression",
    "ruleId": "eslint-interactive/prefer-addition-shorthand",
    "severity": 2,
    "suggestions": [
      {
        "desc": "Use \`val += 1\` instead.",
        "fix": {
          "range": [
            0,
            9,
          ],
          "text": "a += 1",
        },
      },
      {
        "desc": "Use \`val++\` instead.",
        "fix": {
          "range": [
            0,
            9,
          ],
          "text": "a++",
        },
      },
      {
        "desc": "Use \`++val\` instead.",
        "fix": {
          "range": [
            0,
            9,
          ],
          "text": "++a",
        },
      },
    ],
  },
  "suggestions": [
    {
      "desc": "Use \`val += 1\` instead.",
      "fix": {
        "range": [
          0,
          9,
        ],
        "text": "a += 1",
      },
    },
    {
      "desc": "Use \`val++\` instead.",
      "fix": {
        "range": [
          0,
          9,
        ],
        "text": "a++",
      },
    },
    {
      "desc": "Use \`++val\` instead.",
      "fix": {
        "range": [
          0,
          9,
        ],
        "text": "++a",
      },
    },
  ],
}
`;
